spring-kafka/src/main/java/org/springframework/kafka/support/converter/BatchMessagingMessageConverter.java

```diff
@@ -106,6 +141,21 @@ public void setGenerateTimestamp(boolean generateTimestamp) {
 			offsets.add(record.offset());
 			timestampTypes.add(record.timestampType().name());
 			timestamps.add(record.timestamp());
+			if (this.headerMapper != null) {
+				Map<String, Object> converted = new HashMap<>();
+				this.headerMapper.toHeaders(record.headers(), converted);
+				convertedHeaders.add(converted);
+			}
+			else {
+				if (record.headers().iterator().hasNext() && logger.isDebugEnabled() && !logged) {
```

Comment:
Makes sense; this code was also from before (when I warned about dropping headers).